### PR TITLE
refs[TASK][32816]: Handle event Genres Screen

### PR DESCRIPTION
**Pull request by Domhoan**
## Related Tickets
- [#Task 32816: Handle event Genres Screen](https://edu-redmine.sun-asterisk.vn/issues/32816)
## WHAT
- Handle event Genres Screen
 image
image
## HOW
- Create ViewModel
- Bind data to layout
- Create Navigation Graph
- Fix base and module
## WHY
No
## Review Checklist

Category | View Point | Description | Expected Reviewer Answer | Self review | Reviewer2 (name)
--- | --- | --- | --- | --- | ---
Conventions | Does the code follow Sun* coding style and coding conventions? | https://github.com/framgia/coding-standards/tree/master/eng/android | YES |<li>- [ ] yes</li>|<li>- [ ] yes</li>  
Redmine | Does the ticket follow Sun* Redmine working process?  | https://github.com/framgia/Training-Guideline/blob/master/WorkingProcess/redmine/redmine.md| YES |<li>- [ ] yes</li>|<li>- [ ] yes</li>  
Documentation | Is there any incomplete code? If so, should it be removed or flagged with a suitable marker like ‘TODO’? |  | YES |<li>- [ ] yes</li>|<li>- [ ] yes</li>  
## Notes (Optional)
*(Impacted Areas in Application(List features, api, models or services that this PR will affect))*

*(List gem, library third party add new)*

*(Other notes)*

### CODE REVIEW

File: app/src/main/java/com/example/muvi/data/repository/MovieRepositoryType.kt [hunk 1]
```diff
@@ -43,6 +46,26 @@ class MovieRepositoryType(private val remote: MovieDataSource.Remote) : MovieRep
             getMovies(it)
         }
 
+    override fun getMoviesByGenre(genreId: Int, page: Int?): Observable<List<Movie>> {
+        return remote.getMovieByGenre(genreId, page).map {
+            getMoviesType(it).filter { movie ->
+                !movie.poster.isNullOrEmpty()
+            }
```
Comment (thonc-2135) [inline app/src/main/java/com/example/muvi/data/repository/MovieRepositoryType.kt:53]: hình như dùng được takeUnless đó , thử xem cho gọn

File: app/src/main/java/com/example/muvi/data/repository/MovieRepositoryType.kt [hunk 2]
```diff
@@ -43,6 +46,26 @@ class MovieRepositoryType(private val remote: MovieDataSource.Remote) : MovieRep
...
+        Observable.just(GenreUtil.genres)
+
+    private fun getMoviesType(moviesResponse: MovieResponse): List<Movie> =
+        moviesResponse.movies.mapNotNull {
+            try {
+                Movie(it)
+            } catch (exception: IllegalArgumentException) {
+                null
+            }
+        }
```
Comment (thonc-2135) [inline app/src/main/java/com/example/muvi/data/repository/MovieRepositoryType.kt]: hình như trên rx r thì đâu cần phải try catch nhỉ

File: app/src/main/java/com/example/muvi/ui/adpater/MoviesTypeAdapter.kt [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+    ): BaseViewHolder<Movie, ItemMovieBinding> = MovieHolder(
+        ItemMovieBinding.inflate(
+            LayoutInflater.from(parent.context),
+            parent,
+            false
+        ),
+        listener
+    )
+
+    class MovieHolder(private val itemMovieBinding: ItemMovieBinding, listener: (Movie) -> Unit) :
```
Comment (thonc-2135) [inline app/src/main/java/com/example/muvi/ui/adpater/MoviesTypeAdapter.kt]: xuống dòng cho như MoviesTypeAdapter

File: app/src/main/java/com/example/muvi/ui/genres/GenresViewModel.kt [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+    }
+
+    private fun getGenre() {
+        movieRepository.getGenres()
+            .subscribeOn(Schedulers.io())
+            .observeOn(AndroidSchedulers.mainThread())
+            .subscribe(
+                {
+                    _genre.value = it
+                    println(_genre.value)
```
Comment (thonc-2135) [inline app/src/main/java/com/example/muvi/ui/genres/GenresViewModel.kt]: sao phải println thế

File: app/src/main/java/com/example/muvi/ui/genres/GenresViewModel.kt [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+    }
+
+    private fun getData(genreId: Int) {
+        movieRepository.getMoviesByGenre(genreId, ++currentPage)
+            .subscribeOn(Schedulers.io())
+            .observeOn(AndroidSchedulers.mainThread())
+            .subscribe(
+                {
+                    if (currentPage == 1) _movie.value = it as MutableList<Movie>
+                    else _movie.value?.addAll(it)
```
Comment (thonc-2135) [inline app/src/main/java/com/example/muvi/ui/genres/GenresViewModel.kt]: thêm ngoặc kép

File: app/src/main/java/com/example/muvi/ui/genres/MovieTypeFragment.kt [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+
+    override val layoutResource: Int
+        get() = R.layout.fragment_movie_type
+
+    override val viewModel by sharedViewModel<GenresViewModel>()
+
+    private val args: MovieTypeFragmentArgs by navArgs()
+
+    private val moviesTypeAdapter = MoviesTypeAdapter(::onItemClick) {
+        viewModel.getMovies(GenreUtil.getIdGenre(args.title!!))
```
Comment (thonc-2135) [inline app/src/main/java/com/example/muvi/ui/genres/MovieTypeFragment.kt]: null thì sao

File: app/src/main/java/com/example/muvi/base/BaseAdapter.kt [hunk 7]
```diff
@@ -4,11 +4,16 @@ import androidx.recyclerview.widget.ListAdapter
...
-abstract class BaseAdapter<T : GeneralEntity, B : ViewBinding>
-    : ListAdapter<T, BaseViewHolder<T, B>>(BaseDiffUtil<T>()), BindDataAdapter<List<T>> {
+abstract class BaseAdapter<T : GeneralEntity, B : ViewBinding>(
+    private val listener: (T) -> Unit,
+    private val loadMore: () -> Unit
+) : ListAdapter<T, BaseViewHolder<T, B>>(BaseDiffUtil<T>()), BindDataAdapter<List<T>> {
 
     override fun onBindViewHolder(holder: BaseViewHolder<T, B>, position: Int) {
         holder.onBind(getItem(position))
+        if (position == currentList.size - 7) {
```
Comment (thonc-2135) [inline app/src/main/java/com/example/muvi/base/BaseAdapter.kt]: 7 là gì

File: app/src/main/java/com/example/muvi/ui/genres/GenresViewModel.kt (hunk 4 above)
Comment (Domhoan) [reply app/src/main/java/com/example/muvi/ui/genres/GenresViewModel.kt]: hi, mấy hôm trước e debug quên chưa xóa

File: app/src/main/java/com/example/muvi/data/repository/MovieRepositoryType.kt (hunk 1 above)
Comment (Domhoan) [reply app/src/main/java/com/example/muvi/data/repository/MovieRepositoryType.kt:53]: e vừa thử, k được a ạ

**Pull request merged**